### ISSUE TITLE
Incompatibility between .cljs.dev CLJS compiler options file and --repl option ?

### ISSUE THREAD

**Issue by DjebbZ** [bhauman/figwheel-main#8]
Hello,

Given this deps.edn file (complete, not truncated):

```
{:deps      {org.clojure/clojure       {:mvn/version "1.9.0"}
             org.clojure/clojurescript {:mvn/version "1.10.339"}
             com.cerner/clara-rules    {:mvn/version "0.18.0"}
             com.rpl/specter           {:mvn/version "1.1.0"}
             tracking-spec             {:mvn/version "0.1.4-SNAPSHOT"}}
 :paths     ["src"]
 :aliases   {:fig {:extra-deps  {com.bhauman/figwheel-main       {:mvn/version "0.1.4"}}
                   :extra-paths ["target"]
                   :main-opts   ["--main" "figwheel.main" "--build" "dev" "--repl"]}}}
```

When I launch `clojure -A:fig` I noticed that the only `dev.cljs.edn` that will launch a REPL after the build is an almost empty one, with just the :main namespace, like `{:main "trackers.core"}`. Same if I instead use cli option `--compile trackers.core`. If I add any other [CLJS compiler option](https://clojurescript.org/reference/compiler-options) the REPL supposed to launch after the build with the `--repl` option doesn't launch, with no error in the browser console or in the terminal output:

```
#... [truncated, all the normal Figwheel output]
In the cljs.user ns, controls can be called without ns ie. (conns) instead of (figwheel.repl/conns)
    Docs: (doc function-name-here)
    Exit: :cljs/quit
 Results: Stored in vars *1, *2, *3, *e holds last exception object
2018-07-12 16:06:30.763:INFO::main: Logging initialized @19570ms
Opening URL http://localhost:9500
# No ClojureScript version
# No REPL prompt
```

I tried waiting for several minutes, nothing happened.

If I put a complete `dev.cljs.edn` file, all the build settings are honored, notably the `:modules` and `:libs` that I really need, but no REPL is launched. I narrowed it down to the `:main` option being the only one that launches a REPL after removing all other options one by one.

To be sure, I've just re-read again the complete README and Tutorial to make sure I understood things clearly. CLJS compiler options go into .cljs.edn files and Figwheel options in figwheel-main.edn, right ?

End of the (bug?) report !

**Comment by bhauman** [comment]
Using :modules eliminates the :main option.  I’m sure this is the problem you are having.  Using modules eliminates preloads and closure defines.  I will have to come up with a solution for this.

**Comment by DjebbZ** [comment]
This is the output of adding `-pc` to print the config in case it helps, and :main is still here:

```
[Figwheel] :pprint-config true - printing config:
---------------------- Figwheel options ----------------------
{:pprint-config true, :mode :repl, :watch-dirs ("src")}
---------------------- Compiler options ----------------------
{:output-dir "target/public/cljs-out/dev",
 :cache-analysis true,
 :modules
 {:cljs-base {:output-to "target/public/cljs-out/trackers.js"},
  :bing
  {:entries #{trackers.bing.wrapper},
   :output-to "target/public/cljs-out/bing.js",
   :depends-on #{:cljs-base}}},
 :rename-prefix "trkpfx",
 :optimizations :simple,
 :parallel-build true,
 :aot-cache false,
 :output-to "target/public/cljs-out/dev-main.js",
 :asset-path "cljs-out/dev",
 :source-map true,
 :libs
 ["trackers/wt2.js"
  "trackers/google_analytics.js"
  "trackers/facebook.js"
  "trackers/bing/tracker.js"
  "trackers/admo.js"
  "trackers/quantcast.js"
  "trackers/tradelab.js"
  "trackers/google_tag_manager.js"
  "trackers/iadvize.js"
  "trackers/olark.js"
  "trackers/zendesk.js"],
 :compiler-stats true,
 :main "trackers.core"}
```

Are you saying that currently fighwheel-main's REPL doesn't work with :modules ?

**Comment by bhauman** [comment]
In short, yes.  It’s just that the modules preclude the initialization code.

Figwheel main relies on including the :output-to file and it should have content that sets the clojure defines and requires the preloads.  

Modules alters this behavior. There is a solution to this.

**Comment by bhauman** [comment]
To be clear I consider this a bug and I'm planning on fixing it.

But, also, looking at your configuration there doesn't seem to be a reason that you have to use modules during development.

You should consider working without modules during development.

**Comment by DjebbZ** [comment]
Glad to hear you consider this a bug, because I think it is too.

Well I want to ensure lazy loading those
modules works while developing, it's really an important feature for us. I
don't want to test the loading only with a production build without hot
reload a REPL etc.

Are you saying I don't need to change my code which uses cljs.loader if I
just remove the modules option ? If so then yes I could temporarily develop
like this.


Le jeu. 12 juil. 2018 à 19:43, Bruce Hauman <notifications@github.com> a
écrit :

**Comment by bhauman** [comment]
OK I've looked into this and it isn't a bug.

When you use modules you need to provide your own `resources/public/index.html` file that requires the correct module `:output-to`s.

In your case:
<script src="cljs-out/trackers.js"></script>
and then optionally
<script src="cljs-out/bing.js"></script>

As there are several `:output-to`s it is really ambiguous as to what `:output-to`s you want to include on the page.

Now I could make it so that when modules are being compiled that the default figwheel page includes the cljs-base `:output-to` but that would still may be confusing behavior. The usage of Modules is advanced usage, which entails being explicit about how and what you are loading when, I think it is fair that figwheel doesn't/can't handle this case.

**Comment by DjebbZ** [comment]
I re-read this whole conversation several times and tried several things, starting with a slightly modified deps.edn for greater flexibility of trials :

``` edn
{:deps      {org.clojure/clojure       {:mvn/version "1.9.0"}
             org.clojure/clojurescript {:mvn/version "1.10.339"}
             com.cerner/clara-rules    {:mvn/version "0.18.0"}
             com.rpl/specter           {:mvn/version "1.1.0"}
             tracking-spec             {:mvn/version "0.1.4-SNAPSHOT"}}
 :paths     ["src"]
 :aliases   {:fig   {:extra-deps  {com.bhauman/figwheel-main {:mvn/version "0.1.4"}}
                     :extra-paths ["target" "resources"]}
             :build {:main-opts ["--main" "figwheel.main" "--build" "dev" "--repl"]}
             :comp  {:main-opts ["--main" "figwheel.main" "--compile" "trackers.core" "--repl"]}}}
```

1. When running `clojure -A:fig:build` with a proper `resources/public/index.html`, the REPL still doesn't connect. Furthermore, lazy-loading a module from `:modules` fails because the URL is wrong. When I call in the index.html file `trackers.bing.loader.load();`, it's supposed to load the `target/public/cljs-out/bing.js` file from `http://localhost:9500/cljs-out/bing.js`, but the browser shows it's trying to load (path obfuscated) ` http://localhost:9500/cljs-out/dev/<absolute/path/from/the/root/of/my/computer>/target/public/cljs-out/trackers.js`.

2. When running `clojure -A:fig:comp` with the same index.html, I got the same problems that above: no REPL and same wrong path.

3. When running `clojure -A:fig:comp` with the a modified script tag that loads `<script src="cljs-out/main.js></script>"`, I got a working REPL ! Obviously modules loading doesn't work since I did only the specify the main namespace with `--compile`. Here's the content of this new main.js (it appeared only with `:comp` alias) :

``` javascript
var CLOSURE_UNCOMPILED_DEFINES = {"figwheel.repl.connect_url":"ws:\/\/localhost:9500\/figwheel-connect?fwprocess=e0f62a"};
var CLOSURE_NO_DEPS = true;
if(typeof goog == "undefined") document.write('<script src="cljs-out/goog/base.js"></script>');
document.write('<script src="cljs-out/goog/deps.js"></script>');
document.write('<script src="cljs-out/cljs_deps.js"></script>');
document.write('<script>if (typeof goog == "undefined") console.warn("ClojureScript could not load :main, did you forget to specify :asset-path?");</script>');
document.write('<script>goog.require("figwheel.core");</script>');
document.write('<script>goog.require("figwheel.main");</script>');
document.write('<script>goog.require("figwheel.repl.preload");</script>');
document.write('<script>goog.require("devtools.preload");</script>');
document.write('<script>goog.require("figwheel.main.evalback");</script>');
document.write('<script>goog.require("figwheel.main.testing");</script>');
document.write('<script>goog.require("process.env");</script>');
document.write('<script>goog.require("trackers.core");</script>');
```

Clearly it manages to includes all figwheel related stuff.

4. When running `clojure -A:fig:comp` with an added `-pc` flag, I see the configuration included `:preloads`. So I copied it into my `dev.cljs.edn` file and ran it with `clojure -A:fig:comp`, but it failed with an unexpected error:

```
> clojure -A:fig:build
[Figwheel] Compiling build dev to "target/public/cljs-out/dev-main.js"
WARNING: :preloads should only be specified with :none optimizations
WARNING: qualified-keyword? already refers to: #'clojure.core/qualified-keyword? in namespace: clara.rules.dsl, being replaced by: #'clara.rules.dsl/qualified-keyword?
Compile sources, elapsed time: 11845.466691 msecs
WARNING: You required cljs-devtools library in a project which is currently compiled with :optimizations :simple.
         You should remove this library from non-dev builds completely because it impedes dead code elimination.
         The best way is to use :preloads compiler option: https://github.com/binaryage/cljs-devtools/blob/master/docs/installation.md#install-it-via-preloads.
         To silence this warning please set :silence-optimizations-warning config key to true.
         More details: https://github.com/binaryage/cljs-devtools/issues/37.
Compile sources, elapsed time: 3006.034012 msecs
Jul 13, 2018 10:55:35 AM com.google.javascript.jscomp.LoggerErrorManager println
SEVERE: /absolute/path/to/project/target/public/cljs-out/dev/figwheel/core.js:617: ERROR - Syntax error in JS String literal
goog.require(cljs.core.name.call(null,ns_39679),true);
goog.require(cljs.core.name.call(null,ns_39679),true);
^

Jul 13, 2018 10:55:35 AM com.google.javascript.jscomp.LoggerErrorManager println
SEVERE: /absolute/path/to/project/target/public/cljs-out/dev/figwheel/core.js:646: ERROR - Syntax error in JS String literal
goog.require(cljs.core.name.call(null,ns_39699),true);
goog.require(cljs.core.name.call(null,ns_39699),true);
^

Jul 13, 2018 10:55:35 AM com.google.javascript.jscomp.LoggerErrorManager printSummary
WARNING: 2 error(s), 0 warning(s)
ERROR: DEPS_PARSE_ERROR. Syntax error in JS String literal
goog.require(cljs.core.name.call(null,ns_39679),true); at /absolute/path/to/project/target/public/cljs-out/dev/figwheel/core.js line 617 : 0
ERROR: DEPS_PARSE_ERROR. Syntax error in JS String literal
goog.require(cljs.core.name.call(null,ns_39699),true); at /absolute/path/to/project/target/public/cljs-out/dev/figwheel/core.js line 646 : 0
[Figwheel] Failed to compile build dev in 15.655 seconds.
[Figwheel] Compile Exception: Closure compilation failed  
Exception in thread "main" java.lang.Exception: Closure compilation failed
	at cljs.closure$report_failure.invokeStatic(closure.clj:411)
	at cljs.closure$report_failure.invoke(closure.clj:403)
	at cljs.closure$optimize_modules.invokeStatic(closure.clj:1458)
	at cljs.closure$optimize_modules.doInvoke(closure.clj:1409)
	at clojure.lang.RestFn.applyTo(RestFn.java:139)
	at clojure.core$apply.invokeStatic(core.clj:659)
	at clojure.core$apply.invoke(core.clj:652)
	at cljs.closure$build.invokeStatic(closure.clj:2963)
	at cljs.closure$build.invoke(closure.clj:2838)
	at cljs.build.api$build.invokeStatic(api.clj:208)
	at cljs.build.api$build.invoke(api.clj:189)
	at clojure.lang.AFn.applyToHelper(AFn.java:160)
	at clojure.lang.AFn.applyTo(AFn.java:144)
	at clojure.core$apply.invokeStatic(core.clj:657)
	at clojure.core$apply.invoke(core.clj:652)
	at figwheel.main$wrap_with_build_logging$fn__4514$fn__4520.invoke(main.cljc:81)
	at figwheel.main$wrap_with_build_logging$fn__4514.doInvoke(main.cljc:77)
	at clojure.lang.RestFn.invoke(RestFn.java:460)
	at figwheel.main$build.invokeStatic(main.cljc:1158)
	at figwheel.main$build.invoke(main.cljc:1151)
	at figwheel.main$default_compile.invokeStatic(main.cljc:1478)
	at figwheel.main$default_compile.invoke(main.cljc:1457)
	at figwheel.main$build_main_opt.invokeStatic(main.cljc:484)
	at figwheel.main$build_main_opt.invoke(main.cljc:479)
	at cljs.cli$main.invokeStatic(cli.clj:636)
	at cljs.cli$main.doInvoke(cli.clj:625)
	at clojure.lang.RestFn.applyTo(RestFn.java:139)
	at clojure.core$apply.invokeStatic(core.clj:659)
	at clojure.core$apply.invoke(core.clj:652)
	at cljs.main$_main.invokeStatic(main.clj:61)
	at cljs.main$_main.doInvoke(main.clj:52)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.core$apply.invokeStatic(core.clj:657)
	at clojure.core$apply.invoke(core.clj:652)
	at figwheel.main$_main$fn__6548.invoke(main.cljc:1780)
	at clojure.core$with_redefs_fn.invokeStatic(core.clj:7434)
	at clojure.core$with_redefs_fn.invoke(core.clj:7418)
	at figwheel.main$_main.invokeStatic(main.cljc:1778)
	at figwheel.main$_main.doInvoke(main.cljc:1757)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.lang.Var.applyTo(Var.java:702)
	at clojure.core$apply.invokeStatic(core.clj:657)
	at clojure.main$main_opt.invokeStatic(main.clj:317)
	at clojure.main$main_opt.invoke(main.clj:313)
	at clojure.main$main.invokeStatic(main.clj:424)
	at clojure.main$main.doInvoke(main.clj:387)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.lang.Var.applyTo(Var.java:702)
	at clojure.main.main(main.java:37)
```

5. Each time I an `clojure -A:fig:build` I noticed that the path to the compiled JS file displayed by Figwheel is wrong: 

```
[Figwheel] Compiling build dev to "target/public/cljs-out/dev-main.js"
...
[Figwheel] Successfully compiled build dev to "target/public/cljs-out/dev-main.js" in 32.725 seconds.
```

When it should, IMO, display all modules produced. Furthermore this file just doesn't exist.

End of the *long* bug report ! I'm now sure that figwheel and :modules aren't friends (yet). Hopefully, this discussion will drive a better version of figwheel.

Regarding something you said

> The usage of Modules is advanced usage, which entails being explicit about how and what you are loading when, I think it is fair that figwheel doesn't/can't handle this case.

I agree that I have a fairly advanced usage of CLJS in this case, but this doesn't mean I'm an advanced user of *Figwheel*. These are two different things. I would have loved from the start having Figwheel *telling me* that I needed a custom index.html since I was using modules. In my current situation it wouldn't have helped a lot since it's still not working, but at least a helpful message would go a long way.

**Comment by bhauman** [comment]
OMG :optimizations simple.

Figwheel does not work with optimizations simple.

**Comment by bhauman** [comment]
It's important to point out that figwheel is just passing options to the compiler and its doing it's best to inject connection code and make a connection.  If it isn't making that connection we have to debug why.

**Comment by bhauman** [comment]
I've created a separate issue for module related bugs that you have pointed out.

https://github.com/bhauman/figwheel-main/issues/8

**Comment by DjebbZ** [comment]
Ok, from https://clojurescript.org/reference/compiler-options#modules

> Modules are only supported with :simple and :advanced optimizations.

If I don't use `:simple` but instead `:none`/`:whitespace`, I can't use `:modules`. If I can't use `:modules`, I can't write code with lazy-loading using `cljs.loader/load`. Which would mean I would to maintain 2 version of my app, one figwheel-friendly where everything is just `require`d, one for production using `:modules` and lazy-loading. Am I right ? Disillusionment...

**Comment by DjebbZ** [comment]
> I've created a separate issue for module related bugs that you have pointed out.

It's #11

**Comment by bhauman** [comment]
Modules will work in :optimizations :none

**Comment by bhauman** [comment]
https://clojurescript.org/guides/code-splitting

**Comment by bhauman** [comment]
Yes everything works from `:optimizations :none` docs are out of date.

**Comment by DjebbZ** [comment]
Thanks a lot. This deserves a PR on Cljs website, I think I'll do it next
week.

By the way, after a long discussion with the colleagues, we decided that we
don't need modules and code splitting anymore. So without them, figwheel
seems to just works, and rebel-readline-cljs is really nice (I found a bug
though, I will open an issue next week with the details).

Thanks again for all your work and your reactivity.

Le ven. 13 juil. 2018 à 18:04, Bruce Hauman <notifications@github.com> a
écrit :
